benches/macrobenches.rs

```diff
@@ -41,45 +67,143 @@ fn parse_document(doc: &[u8]) -> XmlResult<()> {
     Ok(())
 }
 
-pub fn bench_fully_parse_document(c: &mut Criterion) {
-    let mut group = c.benchmark_group("fully_parse_document");
-
-    let inputs = [
-        // long, mix of attributes and text, not much escaping, mix of attribute lengths, some namespaces
-        ("rpm_primary.xml", RPM_PRIMARY),
-        // long, mix of attributes and text, not much escaping, mix of attribute lengths, some namespaces
-        ("rpm_primary2.xml", RPM_PRIMARY2),
-        // long, mostly medium-length text elements, not much escaping
-        ("rpm_filelists.xml", RPM_FILELISTS),
-        // long, mix of attributes and text, lots of escaping (both entity and char literal), long attributes
-        ("rpm_other.xml", RPM_OTHER),
-        // long, mix of attributes and text, not much escaping, lots of non-ascii characters, lots of namespaces
-        ("libreoffice_document.fodt", LIBREOFFICE_DOCUMENT),
-        // medium length, mostly empty tags, a few short attributes per element, no escaping
-        ("document.xml", DOCUMENT),
-        // medium length, lots of namespaces, no escaping
-        ("test_writer_ident.xml", TEST_WRITER_INDENT),
-        // short, mix of attributes and text, lots of escapes
-        ("sample_1.xml", SAMPLE_1),
-        // medium length, lots of attributes, short attributes, few escapes
-        ("linescore.xml", LINESCORE),
-        // short, lots of namespaces, no escapes
-        ("sample_ns.xml", SAMPLE_NS),
-        // long, few attributes, mix of attribute lengths, escapes in text content
-        ("sample_rss.xml", SAMPLE_RSS),
-        // long, lots of attributes, short attributes, no text, no escapes
-        ("players.xml", PLAYERS),
-    ];
-
-    for (id, data) in inputs.iter() {
+// TODO: use fully normalized attribute values
+fn parse_document_from_bytes(doc: &[u8]) -> XmlResult<()> {
+    let mut r = Reader::from_reader(doc);
+    let mut buf = Vec::new();
+    loop {
+        match r.read_event_into(&mut buf)? {
+            Event::Start(e) | Event::Empty(e) => {
+                for attr in e.attributes() {
+                    criterion::black_box(attr?.decode_and_unescape_value(&r)?);
+                }
+            }
+            Event::Text(e) => {
+                criterion::black_box(e.unescape()?);
+            }
+            Event::CData(e) => {
+                criterion::black_box(e.into_inner());
+            }
+            Event::End(_) => (),
+            Event::Eof => break,
+            _ => (),
+        }
+        buf.clear();
+    }
+    Ok(())
+}
+
+// TODO: use fully normalized attribute values
+fn parse_document_from_str_with_namespaces(doc: &str) -> XmlResult<()> {
+    let mut r = NsReader::from_str(doc);
+    loop {
+        match r.read_resolved_event()? {
+            (_, Event::Start(e) | Event::Empty(e)) => {
```

Comment:
I'm not certain if something more needs to be done to prevent the namespace code from being entirely optimized out.

Reply:
I think you need to send resolved namespace into `black_box`

Reply:
FYI the current API seems to have a slight practical issue:

```
error[E0502]: cannot borrow `r` as immutable because it is also borrowed as mutable
   --> benches/macrobenches.rs:103:74
    |
100 |         match criterion::black_box(r.read_resolved_event()?) {
    |                                    ----------------------- mutable borrow occurs here
...
103 |                     criterion::black_box(attr?.decode_and_unescape_value(&r)?);
    |                                                                          ^^ immutable borrow occurs here
104 |                 }
105 |                 criterion::black_box(resolved_ns);
    |                                      ----------- mutable borrow later used here
```

for 

```
    loop {
        match criterion::black_box(r.read_resolved_event()?) {
            (resolved_ns, Event::Start(e) | Event::Empty(e)) => {
                for attr in e.attributes() {
                    criterion::black_box(attr?.decode_and_unescape_value(&r)?);
                }
                criterion::black_box(resolved_ns);
            }
.....
```

Reply:
Yes, I known. That is the one reason why I've put `Decoder` into events and would prefer to take `Decoder` instead of `&Reader` in attributes until better solution would be developed.

In that case the situation could be resolved, if you consume `resolved_ns` first